Zend/zend_compile.c

```diff
@@ -2447,6 +2452,18 @@ static void zend_emit_return_type_check(
 			return;
 		}
 
+		/* `return` is illegal in a noreturn function */
+		if (ZEND_TYPE_CONTAINS_CODE(type, IS_NORETURN)) {
+			if (implicit) {
+				/* add run-time check in case we do end up returning */
+				zend_emit_op(NULL, ZEND_VERIFY_NORETURN_TYPE, NULL, NULL);
```

Comment:
I still think this should only contain the compilation error (A noreturn function must not return). For `noreturn` functions it would be better if no implicit return is generated at all. You could create a simple opcache test that dumps the opcodes to make sure the opcodes make sense.

Reply:
So there'd be no runtime checks? I'm not sure that matches PHP's normal behaviour. People expect that adding types will mean runtime errors whenever the types are misapplied.

Reply:
The `ZEND_VERIFY_NORETURN_TYPE` is definitely necessary. Currently, your code generates something like this:

```php
function xyz(): noreturn {
    ZEND_VERIFY_NORETURN_TYPE;
    ZEND_RETURN NULL;
}
```

The `ZEND_RETURN` is unnecessary because the `ZEND_VERIFY_NORETURN_TYPE` will always throw. You could make the optimizer understand that `ZEND_VERIFY_NORETURN_TYPE` is terminating so that the return gets removed or you just don't generate it in the first place.

What I'm suggesting is that you check for `noreturn` before calling `zend_emit_final_return(0);` and instead of calling `zend_emit_final_return` you just generate the `ZEND_VERIFY_NORETURN_TYPE`. The error "A noreturn function must not return" can stay in `zend_emit_return_type_check`.